Source/Details/ASCollectionViewLayoutCalculating.h

```diff
@@ -0,0 +1,57 @@
+//
+//  ASCollectionViewLayoutCalculating.h
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 28/2/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <AsyncDisplayKit/ASDimension.h>
+
+@class ASLayout, ASElementMap;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol ASCollectionViewLayoutCalculating <NSObject>
+
+/**
+ * @abstract Return a layout of elements.
+ *
+ * @param size The size that the resulting layout should fit, usually equals to the size of the containing collection view.
+ *
+ * @param map An element map containings all elements to be laid out.
+ *
+ * @discussion This method can be called on multiple theads and thus must be thread-safe (and ideally, stateless).
+ * This method must block the calling thread but can dispatch to other theads to reduce the blocking time.
+ */
+- (ASLayout *)layoutThatFits:(CGSize)size forElementMap:(ASElementMap *)map;
+
+@end
+
+@protocol ASCollectionViewLayoutCalculatorProviding <NSObject>
+
+- (nullable id<ASCollectionViewLayoutCalculating>)collectionViewLayoutCalculator;
+
+- (CGSize)sizeForCalculatingCollectionViewLayout;
+
+@end
+
+@protocol ASCollectionViewLayoutCalculatorResultConsuming <NSObject>
+
+/**
+ * Set a new layout calculated for the given element map.
+ *
+ * @param layout The new layout
+ *
+ * @param calculator The calculator used to calculate this layout
+ *
+ * @param map The element map that was used to calculate the layout
+ *
+ * @note The map might be weakly retained by the consumer for later re-calculations.
+ */
+- (void)setLayout:(ASLayout *)layout calculatedBy:(id<ASCollectionViewLayoutCalculating>)calculator forElementMap:(ASElementMap *)map;
```

Comment:
This would more naturally follow delegate conventions if lead by the calculator, e.g.

layoutCalculator:didFinishLayout:forElementMap: